pkg/lib/indexer/indexer.go

```diff
@@ -696,3 +776,124 @@ func (i ImageIndexer) DeprecateFromIndex(request DeprecateFromIndexRequest) erro
 
 	return nil
 }
+
+func ensureCSVFields(bundleDir string, bundleInfo bundleDirPrefix) error {
+	dirContent, err := ioutil.ReadDir(bundleDir)
+	if err != nil {
+		return fmt.Errorf("error reading bundle directory %s, %v", bundleDir, err)
+	}
+
+	var foundCSV bool
+	for _, f := range dirContent {
+		if f.IsDir() {
+			continue
+		}
+		csvFile, err := os.OpenFile(path.Join(bundleDir, f.Name()), os.O_RDWR, 0)
+		if err != nil {
+			continue
+		}
+		defer csvFile.Close()
+		unstructuredCSV := unstructured.Unstructured{}
+
+		decoder := k8syaml.NewYAMLOrJSONDecoder(csvFile, 30)
+		if err = decoder.Decode(&unstructuredCSV); err != nil {
+			continue
+		}
+
+		if unstructuredCSV.GetKind() != operators.ClusterServiceVersionKind {
+			continue
+		}
+
+		if foundCSV {
+			return fmt.Errorf("more than one ClusterServiceVersion is found in bundle")
+		}
+
+		csv := pregistry.ClusterServiceVersion{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredCSV.UnstructuredContent(),
+			&csv); err != nil {
+			return err
+		}
+		foundCSV = true
+		var modified bool
+		csvReplaces, err := csv.GetReplaces()
+		if err != nil {
+			return err
+		}
+		csvSkips, err := csv.GetSkips()
+		if err != nil {
+			return err
+		}
+		skipMap := make(map[string]struct{})
+		for _, s := range csvSkips {
+			skipMap[s] = struct{}{}
+			if s == bundleInfo.replaces {
+				bundleInfo.replaces = ""
+			}
+		}
+		var updatedSkips bool
+		for s := range bundleInfo.skips {
+			if _, ok := skipMap[s]; !ok {
+				if s == bundleInfo.replaces || s == csvReplaces {
+					// don't add a skips entry already in replaces
+					continue
+				}
+				csvSkips = append(csvSkips, s)
+				skipMap[s] = struct{}{}
+				updatedSkips = true
+			}
+		}
+		if len(bundleInfo.replaces) > 0 && csvReplaces != bundleInfo.replaces {
+			if err := csv.SetReplaces(bundleInfo.replaces); err != nil {
+				return err
+			}
+			if _, ok := skipMap[csvReplaces]; !ok && len(csvReplaces) > 0 {
+				csvSkips = append(csvSkips, csvReplaces)
+				updatedSkips = true
+			}
+			modified = true
+		}
+		if updatedSkips {
+			sort.Strings(csvSkips)
+			if err := csv.SetSkips(csvSkips); err != nil {
+				return err
+			}
+			modified = true
+		}
+
+		if modified {
+			unstructuredCSV, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&csv)
+			if err != nil {
+				return fmt.Errorf("unable to convert modified csv: %v", err)
+			}
+
+			// remove creationTimestamp: null
```

Comment:
aren't these just plain yaml manifests in a bundle? why would there be a creationTimestamp in the metadata of the CSV if that field is set by the api-server

Reply:
[CreationTimestamp](https://github.com/operator-framework/operator-registry/blob/36dc231df42d611494a69fa5b4d1868ce06f2100/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L196) isn't a pointer field on the CSV struct, it gets set to a zero timestamp. This would be exported unless the field was removed.

We can also switch to https://github.com/clarketm/json for the export, that would omit all zero-valued fields.